File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        if(s.size()==0) return "";
        reverse(s.begin(),s.end());
        int n = s.size();
        int l=0,r=0;
        int i=0;
        while(i<n) {
            while(i<n && s[i]==' ') i++;
            if(i==n) break;
            while(i<n && s[i]!=' ') {
                s[r++]=s[i];
                i++;
            }
            reverse(s.begin()+l, s.begin()+r);
            s[r++]=' ';
            l=r;
            i++;
        }
        return s.substr(0,r-1);
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        StringBuilder sb = new StringBuilder();
        int i = s.length()-1;
        while(i>=0) {
            StringBuilder ssb = new StringBuilder();
            while(i>=0 && s.charAt(i)==' ') i--;
            while(i>=0 && s.charAt(i)!=' ') {
                ssb.append(s.charAt(i));
                i--;
            }
            if(ssb.length()>0) {
                if(sb.length()==0) sb.append(ssb.reverse().toString());
                else sb.append(" "+ ssb.reverse().toString());
            }
        }
        return sb.toString();
    }
}